### PR TITLE
refactor: Abstracted registering of new aggregators into a Harvester class that is responsible for starting, stopping, updating all registered aggregators.

### LINKED ISSUE #1981: Provide ability to send common events
### Description
The APM Agent teams are often asked to add support for a new type of data (ex: log events) that will be sent to a new method on the collector. This implementation typically involves duplicating a set of logic so that the new data type has its own record_<data type> function that creates a record of the new data type, list/queue/aggregator for the data to be sent, configuration for tuning how much information to hold before sampling, configuration for how often to harvest the data, and logic to actually harvest the new list/queue/aggregator data on the correct intervals. 

Instead of cloning the logic required to support a new data type sent to a new collector method (build the list/queue/aggregator, configuration, harvester, etc) the proposal is to build application logic that would allow defining a new type of custom event with its own configuration, list/queue/aggregator, and harvester.  It is a subtle distinction in that most of the effort is the same. The key difference is that this agent implementation should allow agents to support a new category of event in the future that has it’s own sampling rate with minimal future changes to the agent required to support that new category.

### Acceptance Criteria
As an agent engineer, it should be clear on how to add a new aggregator and all the necessary logic for it to function

### Additional context
 This is a recent-ish [PR](https://github.com/newrelic/node-newrelic/pull/1147/files) that implemented a new aggregator and all the necessary plumbing to enqueue and send a new type of events.

### PR DESCRIPTION
**Pull request by bizob2828**
## Description

Sorry for the large PR 😬.  This PR removes all the places that are necessary to add certain calls when adding a new aggregator.  This introduces a harvester class that is responsible for tracking all aggregators.  Then depending on stages in the agent it will start, stop, update, and clear the aggregators. To do that I had to update the signature of every aggregator to pass in the harvester instance because registering of an aggregator is done at the construction via `harvest.add`. Also I added an evaluator config to determine when an aggregator is enabled based on relevant configuration. Previously this was done in `startAggregators` and was easy to miss when adding a new aggregator.

This achieves 1 part of the making it easier to add a new aggregator for the future.  I deferred having a common agent method for enqueuing events to another story.  The idea behind that is instead of calling the aggregators directly we will have a generic `agent.addEvent` method that takes `id` and `event` and `priority`.  So right now it's

```js
agent.customEventAggregator.add([{ type, timestamp }, event])
```

It'd be

```js
agent.addEvent(CUSTOM_EVENT, [{ type, timestamp}, event])
```

This functionality will come next. Then we can add the LLM aggregator with its relevant config(still being discussed what the values for config are).

Lastly, I added some missing coverage or made more direct tests around behaviors as it was very difficult to see where things were being tested.  

## Related Issues
Closes #1981

### CODE REVIEW

File: lib/aggregators/base-aggregator.js [hunk 1]
```diff
@@ -9,18 +9,32 @@ const EventEmitter = require('events').EventEmitter
...
     this.defaultLimit = this.limit = opts.limit
     this.runId = opts.runId
     this.isAsync = opts.isAsync || false
+    // function to pass in to determine if we can start a given aggregator
+    this.isEnabled =
+      opts.enabled ||
+      function defaultEnabled() {
+        return true
+      }
+    this.enabled = this.isEnabled(opts.config)
```
Comment (jsumners-nr) [inline lib/aggregators/base-aggregator.js:25]: I've been reading this over and over for a minute or two. It's giving me some sort of "uhhh" feeling. Is this just some backward compatibility thing?

File: lib/collector/api.js [hunk 2]
```diff
@@ -70,26 +70,38 @@ function CollectorAPI(agent) {
...
+CollectorAPI.prototype.send = function send(method, data, callback) {
+  if (!callback) {
+    this._throwCallbackError()
+  }
+  if (!data) {
+    callback(new TypeError(`must pass data for ${method} to send`))
+    return
+  }
+
+  this._sendData(this._methods[method], data, callback)
```
Comment (jsumners-nr) [inline lib/collector/api.js:102]: For other people reviewing this: the changes in lines 73-86 are tied to `this._methods`. This PR is simplifying that mapping.

File: lib/harvester.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2024 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+module.exports = class Harvester {
```
Comment (jsumners-nr) [inline lib/harvester.js:8]: I'd suggest a jsdoc block that describes what this container is, what it replaces, and just the general overview of how it is intended to work.

**Review by jsumners-nr — DISMISSED** [review]
I haven't gone over every line since it's mostly just massaging things into the new signatures. I have one piece of feedback, but overall 👍

File: lib/aggregators/base-aggregator.js (hunk 1 above)
Comment (bizob2828) [reply lib/aggregators/base-aggregator.js:25]: so an aggregator can be enabled but based on server side configuration updates become disabled.  This centralizes the configuration bits to determine that.  

This used to live [here](https://github.com/newrelic/node-newrelic/blob/main/lib/agent.js#L310), [here](https://github.com/newrelic/node-newrelic/blob/main/lib/agent.js#L451), and [here](https://github.com/newrelic/node-newrelic/blob/main/lib/agent.js#L485).  It has to be a function to evaluate at the config can change in reconfigure, I'm totally open for better solutions, it was the best i could think of at the time

**Review by jsumners-nr — APPROVED** [review]

**Pull request merged**